src/client/datascience/jupyter/kernels/kernelConnectionWrapper.ts

```diff
@@ -44,84 +44,128 @@ export class KernelConnectionWrapper implements Kernel.IKernelConnection {
     public readonly unhandledMessage = new Signal<this, IMessage<MessageType>>(this);
     public readonly anyMessage = new Signal<this, Kernel.IAnyMessageArgs>(this);
     public get serverSettings() {
-        return this.kernelConnection.serverSettings;
+        return (this.kernelConnection || this._previousKernelConnection).serverSettings;
     }
     public readonly disposed = new Signal<this, void>(this);
+    /**
+     * Use `kernelConnection` to access the value as its not a constant (can change over time).
+     * E.g. when restarting kernels or the like.
+     */
+    private _kernelConnection!: Kernel.IKernelConnection;
+    private readonly _previousKernelConnection: Kernel.IKernelConnection;
+    // private _isRestarting?: boolean;
+    private get kernelConnection(): undefined | Kernel.IKernelConnection {
+        if (this.kernel.session?.kernel === this._kernelConnection) {
+            return this._kernelConnection;
+        }
+        this.stopHandlingKernelMessages(this._kernelConnection);
+        if (this.kernel.session?.kernel) {
+            this.startHandleKernelMessages(this.kernel.session.kernel);
+            return this._kernelConnection;
+        }
+    }
+    private getKernelConnection(): Kernel.IKernelConnection {
```

Comment:
What's the difference between this and just ```kernelConnection()```. Seems weird to have both?

Reply:
This will throw an exception if kernelConnection is undefined, and when using the accessor `this.kernelConnection` we don't want exceptions to be thrown.

I.e. the property kernelConnection can be undefined. however the return value of getKernelConnection can never be undefined.

Reply:
Still seems weird. Maybe name them something different? Like ```possibleKernelConnection``` and ```kernelConnection```?

Reply:
Will rename to `getPossibleKernelConnection`
The problem is, e.g. when calling `executeRequest` at that point a kernel connection is required, and we cannot call it without a kernel conneciton.
If 3rd part calls it in such a state (disposed state, then we must handle it that case and throw an exception)
Hence the check to see if its undefined & throw an error.

I could optionally rename the method to `getKernelConnectionAndThrowIfNotFound` to make it more obivous.